// vd/node/style.js
"use strict";

var _slice = Array.prototype.slice;
var _toArray = function (arr) {
  return Array.isArray(arr) ? arr : Array.from(arr);
};

var _inherits = function (child, parent) {
  child.prototype = Object.create(parent && parent.prototype, {
    constructor: {
      value: child,
      enumerable: false,
      writable: true,
      configurable: true
    }
  });
  if (parent) child.__proto__ = parent;
};

var _interopRequire = function (obj) {
  return obj && (obj["default"] || obj);
};

var Sheet = _interopRequire(require("vcss"));

var Text = _interopRequire(require("./text"));

var Element = _interopRequire(require("./element"));

var Style = function Style() {
  var _this = this;
  Element.call(this, "style");

  this.sheet = new Sheet();

  {
    var text = new Text();
    text.value = function () {
      return _this.sheet.serialize();
    };
    this.children.push(text);
  }
};

_inherits(Style, Element);

Style.prototype.add = function () {
  var props = _slice.call(arguments);

  this.sheet.add.apply(this.sheet, _toArray(props));
  return this;
};

Style.prototype.remove = function (dec) {
  this.sheet.remove(dec);
  return this;
};

Style.prototype.anim = function (name, frames) {
  return this.sheet.anim(name, frames);
};

Style.prototype.media = function (params) {
  return this.sheet.media(params);
};

module.exports = Style;
